File: 99-recover-binary-search-tree/99-recover-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void recoverTree(TreeNode* root) {
        TreeNode* node = root;
        TreeNode* pred = NULL;
        TreeNode* predecessor = NULL;
        TreeNode* x = NULL;
        TreeNode* y = NULL;
        while(node){
            if(node->left){
                predecessor = node->left;
                while(predecessor->right and predecessor->right != node){
                    predecessor = predecessor->right;
                }
                if(predecessor->right == NULL){
                    predecessor->right = node;
                    node = node->left;
                }else{
                    if(pred and node->val < pred->val){
                        y = node;
                        if(x == NULL) x = pred;
                    }
                    pred = node;
                    predecessor->right = NULL;
                    node = node->right;
                }
            }else{
                if(pred and node->val < pred->val){
                    y = node;
                    if(x == NULL) x = pred;
                }
                pred = node;
                node = node->right;
            }
        }
        swap(x->val,y->val);
    }
};

File: 99-recover-binary-search-tree/99-recover-binary-search-tree.java
class Solution {
  public void swap(TreeNode a, TreeNode b) {
    int tmp = a.val;
    a.val = b.val;
    b.val = tmp;
  }

  public void recoverTree(TreeNode root) {
    // predecessor is a Morris predecessor. 
    // In the 'loop' cases it could be equal to the node itself predecessor == root.
    // pred is a 'true' predecessor, 
    // the previous node in the inorder traversal.
    TreeNode x = null, y = null, pred = null, predecessor = null;

    while (root != null) {
      // If there is a left child
      // then compute the predecessor.
      // If there is no link predecessor.right = root --> set it.
      // If there is a link predecessor.right = root --> break it.
      if (root.left != null) {
        // Predecessor node is one step left 
        // and then right till you can.
        predecessor = root.left;
        while (predecessor.right != null && predecessor.right != root)
          predecessor = predecessor.right;

        // set link predecessor.right = root
        // and go to explore left subtree
        if (predecessor.right == null) {
          predecessor.right = root;
          root = root.left;
        }
        // break link predecessor.right = root
        // link is broken : time to change subtree and go right
        else {
          // check for the swapped nodes
          if (pred != null && root.val < pred.val) {
            y = root;
            if (x == null) x = pred;
          }
          pred = root;

          predecessor.right = null;
          root = root.right;
        }
      }
      // If there is no left child
      // then just go right.
      else {
        // check for the swapped nodes
        if (pred != null && root.val < pred.val) {
          y = root;
          if (x == null) x = pred;
        }
        pred = root;

        root = root.right;
      }
    }
    swap(x, y);
  }
}